cgen error (dump): undefined symbol 'cannot convert 'struct main__Aa' to 'struct main__Bb'

### Describe the bug

```bash
==================
/tmp/v_1000/16691.12100382184385958814.tmp.c:12242: error: cannot convert 'struct main__Aa' to 'struct main__Bb'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Expected Behavior

no error

### Current Behavior

error

### Reproduction Steps

```v
module main

struct Aa {
	sub AliasType
}

type AliasType = Bb

struct Bb {
	a int
}

fn encode_struct[U](val U) {
	dump(val) // this not work
	println(val)  // this work
	$for field in U.fields {
		encode_struct(val.$(field.name))
	}
}

fn main() {
	aa := Aa{}

	encode_struct(aa)
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.2 afdb097

### Environment details (OS name and version, etc.)

Ubuntu

--- a/vlib/v/gen/c/dumpexpr.v
+++ b/vlib/v/gen/c/dumpexpr.v
@@ -19,8 +19,7 @@ fn (mut g Gen) dump_expr(node ast.DumpExpr) {
 		// generic func with recursion rewrite node.expr_type
 		if node.expr is ast.Ident {
 			// var
-			if (node.expr as ast.Ident).info is ast.IdentVar
-				&& (node.expr as ast.Ident).language == .v {
+			if (node.expr as ast.Ident).info is ast.IdentVar && (node.expr as ast.Ident).language == .v {
 				name = g.typ(g.unwrap_generic((node.expr as ast.Ident).info.typ))
 			}
 		}

--- a/vlib/v/gen/c/dumpexpr.v
+++ b/vlib/v/gen/c/dumpexpr.v
@@ -20,7 +20,8 @@ fn (mut g Gen) dump_expr(node ast.DumpExpr) {
 		if node.expr is ast.Ident {
 			// var
 			if (node.expr as ast.Ident).info is ast.IdentVar && (node.expr as ast.Ident).language == .v {
-				name = g.typ(g.unwrap_generic((node.expr as ast.Ident).info.typ)).replace('*', '')
+				name = g.typ(g.unwrap_generic((node.expr as ast.Ident).info.typ)).replace('*',
+					'')
 			}
 		}
 	}

--- a/vlib/v/gen/c/dumpexpr.v
+++ b/vlib/v/gen/c/dumpexpr.v
@@ -20,7 +20,7 @@ fn (mut g Gen) dump_expr(node ast.DumpExpr) {
 		if node.expr is ast.Ident {
 			// var
 			if (node.expr as ast.Ident).info is ast.IdentVar && (node.expr as ast.Ident).language == .v {
-				name = g.typ(g.unwrap_generic((node.expr as ast.Ident).info.typ))
+				name = g.typ(g.unwrap_generic((node.expr as ast.Ident).info.typ)).replace('*', '')
 			}
 		}
 	}

--- a/vlib/v/gen/c/dumpexpr.v
+++ b/vlib/v/gen/c/dumpexpr.v
@@ -15,6 +15,16 @@ fn (mut g Gen) dump_expr(node ast.DumpExpr) {
 	mut name := node.cname
 	mut expr_type := node.expr_type
 
+	if g.cur_fn != unsafe { nil } && g.cur_fn.generic_names.len > 0 {
+		// generic func with recursion rewrite node.expr_type
+		if node.expr is ast.Ident {
+			// var
+			if (node.expr as ast.Ident).info is ast.IdentVar
+				&& (node.expr as ast.Ident).language == .v {
+				name = g.typ(g.unwrap_generic((node.expr as ast.Ident).info.typ))
+			}
+		}
+	}
 	// var.${field.name}
 	if node.expr is ast.ComptimeSelector {
 		selector := node.expr as ast.ComptimeSelector

--- a/vlib/v/tests/dump_on_generic_func_test.v
+++ b/vlib/v/tests/dump_on_generic_func_test.v
@@ -1,0 +1,25 @@
+interface Refresher {
+	refresh()
+}
+
+struct Dummy {}
+
+fn (d Dummy) refresh() {}
+
+struct Source {
+mut:
+	refresher Refresher = Dummy{}
+}
+
+struct App {
+mut:
+	src Source
+}
+
+fn test_struct_init_with_interface_field() {
+	mut app := &App{}
+	app.src = Source{}
+
+	println(app)
+	assert '${app}'.contains('refresher: Refresher(Dummy{})')
+}

--- a/vlib/v/tests/dump_on_generic_func_test.v
+++ b/vlib/v/tests/dump_on_generic_func_test.v
@@ -1,25 +1,24 @@
-interface Refresher {
-	refresh()
+struct Aa {
+	sub AliasType
 }
 
-struct Dummy {}
+type AliasType = Bb
 
-fn (d Dummy) refresh() {}
-
-struct Source {
-mut:
-	refresher Refresher = Dummy{}
+struct Bb {
+	a int
 }
 
-struct App {
-mut:
-	src Source
+fn encode_struct[U](val U) {
+	dump(val)
+	println(val)
+	$for field in U.fields {
+		encode_struct(val.$(field.name))
+	}
 }
 
-fn test_struct_init_with_interface_field() {
-	mut app := &App{}
-	app.src = Source{}
+fn test_main() {
+	aa := Aa{}
 
-	println(app)
-	assert '${app}'.contains('refresher: Refresher(Dummy{})')
+	encode_struct(aa)
+	assert true
 }

